spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/route/CachingRouteDefinitionLocator.java

```diff
@@ -61,7 +68,9 @@ public CachingRouteDefinitionLocator(RouteDefinitionLocator delegate) {
 
 	@Override
 	public void onApplicationEvent(RefreshRoutesEvent event) {
-		refresh();
+		fetch().materialize().collect(Collectors.toList())
+				.map(routes -> cache.put(CACHE_KEY, routes))
+				.subscribeOn(Schedulers.elastic()).subscribe();
```

Comment:
why does it subscribe on the elastic scheduler?

Reply:
Do you have any recommendations on what it should subscribe? Currently the method is void and can not block (as it is used directly in a WebFlux) controller. Therefore I just choose the `elastic` scheduler as I was not sure if a `Schedulers#single()` is sufficient.

Reply:
`.subscribe` does not block, plus `subscribeOn` affects only upstream operators (`fetch`, in your case, which should already be non-blocking)

Reply:
Exactly what I try to achieve, run the `fetch` on a background thread, don't block the calling thread. If it fails `¯\_(ツ)_/¯`, the interface does not offer any signalling.

Reply:
If a method returns a reactive type (Mono/Flux), it kinda says that it is non-blocking already, that's the contract. It is recommended to use `subscribeOn` only when you create the publisher, because that's your responsibility to return a Publisher that will not block the calling thread

Reply:
Ok, got it. So just subscribing should be sufficient enough. Will change the PR.

Reply:
After removing the `Schedulers.elastic()` I observed the same blocking behaviour I tried to mitigate. The issue is in `org.springframework.cloud.gateway.discovery.DiscoveryClientRouteDefinitionLocator#getRouteDefinitions` which calls `discoveryClient.getServices()`. `discoveryClient.getServices()` is a blocking call and imho not adapted properly. I'll try to fix that with this PR.

Could you have a look at the changes?